File: leetcode/src/test/java/com/giwankim/leetcode/GroupAnagramsTest.java
package com.giwankim.leetcode;

import static org.assertj.core.api.Assertions.assertThat;

import java.util.List;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.ValueSource;

class GroupAnagramsTest {
  GroupAnagrams sut = new GroupAnagrams();

  @ParameterizedTest
  @ValueSource(strings = {"", "a"})
  void singleton(String str) {
    List<List<String>> actual = sut.groupAnagrams(new String[] {str});
    assertThat(actual).isEqualTo(List.of(List.of(str)));
  }

  @Test
  void groupAnagrams() {
    String[] strs = {"eat", "tea", "tan", "ate", "nat", "bat"};
    List<List<String>> expected =
        List.of(List.of("bat"), List.of("nat", "tan"), List.of("ate", "eat", "tea"));

    List<List<String>> actual = sut.groupAnagrams(strs);

    actual.forEach(l -> l.sort(null));
    assertThat(actual).containsExactlyInAnyOrderElementsOf(expected);
  }
}

File: leetcode-kt/src/test/kotlin/com/giwankim/leetcode/GroupAnagramsTest.kt
package com.giwankim.leetcode

import io.kotest.core.spec.style.FunSpec
import io.kotest.datatest.withTests
import io.kotest.matchers.collections.shouldContainExactlyInAnyOrder

class GroupAnagramsTest :
    FunSpec({
        val sut = GroupAnagrams()

        context("group anagrams") {
            withTests(
                nameFn = { (strs, expected) -> "strs=${strs.contentToString()}, expected=$expected" },
                GroupAnagramsCase(
                    arrayOf("eat", "tea", "tan", "ate", "nat", "bat"),
                    listOf(listOf("bat"), listOf("nat", "tan"), listOf("ate", "eat", "tea")),
                ),
                GroupAnagramsCase(arrayOf(""), listOf(listOf(""))),
                GroupAnagramsCase(arrayOf("a"), listOf(listOf("a"))),
            ) { (strs, expected) ->
                val actual = sut.groupAnagrams(strs).map { it.sorted() }
                actual shouldContainExactlyInAnyOrder expected.map { it.sorted() }
            }
        }
    })

private data class GroupAnagramsCase(
    val strs: Array<String>,
    val expected: List<List<String>>,
)
